system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -6,9 +6,92 @@ const TopSitesEdit = require("./TopSitesEdit");
 const {TopSite, TopSitePlaceholder} = require("./TopSite");
 const CollapsibleSection = require("content-src/components/CollapsibleSection/CollapsibleSection");
 const ComponentPerfTimer = require("content-src/components/ComponentPerfTimer/ComponentPerfTimer");
+const {actionCreators: ac} = require("common/Actions.jsm");
+
+const {MIN_RICH_FAVICON_SIZE, MIN_CORNER_FAVICON_SIZE} = require("./TopSitesConstants");
+
+/**
+ * Select TopSite image and favicon based on available properties.
+ * @param link Entry in TopSites.
+ */
+function selectTopSiteIcon(link) {
+  const {tippyTopIcon, faviconSize} = link;
+  let imageClassName;
+  let imageStyle;
+  let smallFaviconStyle;
+  let showSmallFavicon = false;
+  let smallFaviconFallback = false;
+  if (tippyTopIcon || faviconSize >= MIN_RICH_FAVICON_SIZE) {
+    // styles and class names for top sites with rich icons
+    imageClassName = "top-site-icon rich-icon";
+    imageStyle = {
+      backgroundColor: link.backgroundColor,
+      backgroundImage: `url(${tippyTopIcon || link.favicon})`
+    };
+  } else {
+    // styles and class names for top sites with screenshot + small icon in top left corner
+    imageClassName = `screenshot${link.screenshot ? " active" : ""}`;
+    imageStyle = {backgroundImage: link.screenshot ? `url(${link.screenshot})` : "none"};
+
+    // only show a favicon in top left if it's greater than 16x16
+    if (faviconSize >= MIN_CORNER_FAVICON_SIZE) {
+      showSmallFavicon = true;
+      smallFaviconStyle = {backgroundImage:  `url(${link.favicon})`};
+    } else if (link.screenshot) {
+      // Don't show a small favicon if size is too small, because that
+      // would result in two fallback icons.
+      showSmallFavicon = true;
+      smallFaviconFallback = true;
+    }
+  }
+
+  return Object.assign({}, link, {
+    imageClassName,
+    imageStyle,
+    showSmallFavicon,
+    smallFaviconStyle,
+    smallFaviconFallback
+  });
+}
+
+/**
+ * Reducer that iterates through TopSites and counts types of images.
+ * @param acc Accumulator for reducer.
+ * @param topsite Entry in TopSites.
+ */
+function countTopSitesIconsTypes(acc, topsite) {
+  const {
+    screenshot, showSmallFavicon,
+    smallFaviconFallback, tippyTopIcon
+  } = topsite;
+
+  if (screenshot && smallFaviconFallback) {
+    acc.screenshot++;
+  } else if (screenshot && showSmallFavicon) {
+    acc.screenshot_with_icon++;
+  } else if (tippyTopIcon) {
+    acc.tippytop++;
+  } else {
+    acc.rich_icon++;
+  }
+
+  return acc;
+}
 
 const TopSites = props => {
-  const realTopSites = props.TopSites.rows.slice(0, props.TopSitesCount);
+  const realTopSites = props.TopSites.rows.slice(0, props.TopSitesCount)
+                        .map(selectTopSiteIcon);
+
+  const topSitesIconsStats = realTopSites.reduce(countTopSitesIconsTypes, {
```

Comment:
Personally I would go with `countTopSitesIconsTypes` just taking `realTopSites` as an argument and add this `.reduce`/initial state for the accumulator in that function